### DISCUSSION TITLE
filterCurrent on TenantCollection

### DISCUSSION THREAD

**Discussion by chris-ware** [spatie/laravel-multitenancy#266] [Ideas]
Would adding a `filterCurrent` method onto TenantCollection be something you'd consider? I'd be happy to take a look at adding a PR for this, looks like most of the mechanics are already there for it in TenantCollection itself. I'd maybe struggle a bit with the test case, but I'm sure it could be tweaked before merging if needed, unless the awesome team at Spatie want to do this yourselves.

**Comment by masterix21** [comment]
Hi @chris-ware, could you explain better the feature suggested? Can’t you use a filter by id?

**Reply by chris-ware** [reply]
I mean the filter method from the base Laravel collections. In a project I'm working with, I have need to take all tenants and filter them down based on a callback. Currently, I use Tenant::all()->filter() with a makeCurrent and then the callback. An idea I've had since would be to also do the opposite and allow for reject too.

**Reply by masterix21** [reply]
Can't you use the `eachCurrent` method? 

Please post an example of that, because I think that I didn't understand you.

**Reply by chris-ware** [reply]
At the moment, I have to do something like the following:
```php
Tenant::all()->filter(function(Tenant $tenant) {  
    $tenant->makeCurrent();  
    return $tenant->isValid();  
});
```  

This idea isn't a necessity as the code works, it's just a layer of syntactic sugar, so to speak. With a filterCurrent() method on TenantCollection, I could achieve the following:
```php
Tenant::all()->filterCurrent(fn(Tenant $tenant) => $tenant->isValid());
```

**Reply by masterix21** [reply]
Please keep in mind that Laravel Collection is a Macroable class: you can extend it.

I'm not sure if a `filterCurrent` method is necessary. I'd like to hear the @freekmurze opinion about that.

**Reply by freekmurze** [reply]
If the code is simple and well tested, I’d accept a PR that adds that to the package.

**Reply by chris-ware** [reply]
@freekmurze I think I can do that easily following along the same code patterns as the eachCurrent and mapCurrent methods. I'll have a go and send a PR in once I'm done.

**Reply by masterix21** [reply]
@freekmurze, isn't better to use the `__call` magic method to handle all methods using the prefix `Current`? It will solve any further implementation.

**Reply by freekmurze** [reply]
I'm thinking that with `__call` we'll lose autocompletion in IDE's. Let's add this `filterCurrent` now. If any other methods get requests we could opt to refactor them all to `__call`

**Reply by chris-ware** [reply]
I agree, I also think only some methods would need a Current variant in the tenant collection. I would exect that was the intention anyway with the TenantCollection. It also means that if new methods are added to the base Laravel Collection, it wouldn't need testing and new version constraints to ensure it works.
